File: core/src/main/java/com/wind/core/filter/Filter.java
package com.wind.core.filter;

import java.util.concurrent.Future;

/*
* 核心设计模式 -> https://twitter.github.io/finagle/guide/ServicesAndFilters.html
* https://monkey.org/~marius/funsrv.pdf
* */
public interface Filter<ReqIn, RepOut, ReqOut, RepIn> {
	Future<RepOut> apply(ReqIn req, Service<ReqOut, RepIn> service);

	default Service<ReqIn, RepOut> then(Service<ReqOut, RepIn> service) {
		return request -> apply(request, service);
	}

	default <Req, Rep> Filter<ReqIn, RepOut, Req, Rep> andThen(Filter<ReqOut, RepIn, Req, Rep> next) {
		return (request, service) -> apply(request, next.then(service));
	}
}

File: core/src/main/java/com/wind/core/filter/filter.js
/*
* filter: (req, service) => rep
* > filter & filter : filter
* > filter & service : service
*
* >> filter & filter & service : service
* */

function Filter(flow) {
  return (channel => ({
      and: next => Filter((req, srv) => channel(req, next.then(srv))),
      then: srv => req => channel(req, srv)
  }))(flow);
}

function FilterOrigin(flow) {
  class _Inner {
    constructor(ch) {
      this._ch = ch;
    }
    and = next => {
      const self = this;
      return (new FilterOrigin((req, srv) => self._ch(req, next.then(srv))));
    };
    then = req => {
      const self = this;
      return (self._ch(req, srv));
    };
  };

  return new _Inner(flow);
}
